File: 2023/day-01/main2.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

var digitMap = map[string]int{
	"one":   1,
	"two":   2,
	"three": 3,
	"four":  4,
	"five":  5,
	"six":   6,
	"seven": 7,
	"eight": 8,
	"nine":  9,
}

func findCalibrationValuePart2(line string) int {
	// Create a slice to store the indices and values of all digits found
	type digit struct {
		value int
		index int
	}
	digits := []digit{}

	for i := 0; i < len(line); i++ {
		// Check for numeric digits
		if charDigit := line[i] - '0'; charDigit >= 0 && charDigit <= 9 {
			digits = append(digits, digit{int(charDigit), i})
		} else {
			// Check for spelled-out digits
			for word, value := range digitMap {
				if strings.HasPrefix(line[i:], word) {
					digits = append(digits, digit{value, i})
				}
			}
		}
	}

	// Sort the digits by their index to get the first and last
	sort.Slice(digits, func(i, j int) bool {
		return digits[i].index < digits[j].index
	})

	// If any digits were found, return the combination of the first and last
	if len(digits) > 0 {
		return digits[0].value*10 + digits[len(digits)-1].value
	}

	return 0 // If no digits are found
}

func calculateTotalPart2(filename string) (int, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return 0, err
	}

	lines := strings.Split(string(data), "\n")
	total := 0
	for _, line := range lines {
		if line != "" {
			total += findCalibrationValuePart2(line)
		}
	}
	return total, nil
}

func main() {
	filename := "input.txt" // Replace with your actual input file

	total, err := calculateTotalPart2(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	fmt.Println("Sum of all calibration values:", total)
}

File: 2023/day-01/main2.lua
-- Not Working
local digitMap = {
    one = 1,
    two = 2,
    three = 3,
    four = 4,
    five = 5,
    six = 6,
    seven = 7,
    eight = 8,
    nine = 9,
}

--function findCalibrationValuePart2(line)
--    local digits = {}
--
--    -- Construct a pattern string from digitMap keys
--    local digitPatterns = {}
--    for word, _ in pairs(digitMap) do
--        table.insert(digitPatterns, word)
--    end
--    local digitPatternString = table.concat(digitPatterns, "|")
--
--    -- Regular expression with lookahead to prevent overlaps
--    local digitPattern = "(?=(" .. digitPatternString .. "|%d))"
--
--    for match in string.gmatch(line, digitPattern) do
--        local valueStr = match
--        local value = tonumber(valueStr) or digitMap[valueStr]
--        table.insert(digits, { value = value, index = match:find(valueStr) })
--    end
--
--    table.sort(digits, function(a, b) return a.index < b.index end)
--
--    if #digits > 0 then
--        return digits[1].value * 10 + digits[#digits].value
--    else
--        return 0
--    end
--end

--function findCalibrationValuePart2(line)
--    local digits = {}
--
--    local digitPatterns = {}
--    for word, _ in pairs(digitMap) do
--        table.insert(digitPatterns, word)
--    end
--    local digitPatternString = table.concat(digitPatterns, "|")
--
--    local digitPattern = "(?=(" .. digitPatternString .. "|%d))"
--
--    -- Capture and iterate over the sub-matches within the lookahead group
--    for match, valueStr in string.gmatch(line, digitPattern) do
--        local value = tonumber(valueStr) or digitMap[valueStr]
--        table.insert(digits, { value = value, index = match:find(valueStr) })
--    end
--
--    table.sort(digits, function(a, b) return a.index < b.index end)
--
--    if #digits > 0 then
--        return digits[1].value * 10 + digits[#digits].value
--    else
--        return 0
--    end
--end

--function findCalibrationValuePart2(line)
--    local digits = {}
--    local pattern = "(%d|one|two|three|four|five|six|seven|eight|nine)"
--
--    for numberStr, index in line:gmatch(pattern) do
--        local value = tonumber(numberStr) or digitMap[numberStr]
--        table.insert(digits, { value = value, index = index })
--    end
--
--    table.sort(digits, function(a, b) return a.index < b.index end)
--
--    if #digits > 0 then
--        return digits[1].value * 10 + digits[#digits].value
--    else
--        return 0
--    end
--end

function findCalibrationValuePart2(line)
    local digits = {}
    local pattern = "(one|two|three|four|five|six|seven|eight|nine|%d)"

    -- Capture and iterate over the sub-matches within the lookahead group
    for match in string.gmatch(line, pattern) do
        local value = tonumber(match) or digitMap[match]
        table.insert(digits, { value = value, index = line:find(match) })
    end

    table.sort(digits, function(a, b) return a.index < b.index end)

    -- Update logic for 0 digit case
    if #digits > 0 then
        return digits[1].value * 10 + digits[#digits].value
    else
        return 0
    end
end

function calculateTotalPart2(filename)
    local file = io.open(filename, "r")
    if not file then
        error("Error reading file: " .. filename)
    end

    local total = 0
    for line in file:lines() do
        if line ~= "" then
            total = total + findCalibrationValuePart2(line)
        end
    end

    file:close()
    return total
end

local filename = "input.txt"  -- Replace with your actual input file

local total, err = calculateTotalPart2(filename)
if err then
    print(err)
else
    print("Sum of all calibration values:", total)
end
